### ISSUE TITLE
http-cookie-agent currupts requests when keepAlive is true.

### ISSUE THREAD

**Issue by everett1992** [3846masa/http-cookie-agent#65]
## Describe the bug

http-cookie-agent can corrupt requests, causing http parsing errors on the server, when using `keepAlive: true`. 

## To Reproduce
I see this issue consistently when making POST requests with a cookie for the domain.
I included a test script and a server that prints invalid http requests.
 
When I run the test script it errors after three requests 
```
error 2 Request failed with status code 400
```

The server prints something like this after the first reused connection.
```
Parse Error: Invalid method encountered {
  packet: 'POST /1 HTTP/1.1\r\n' +
    'Accept: application/json, text/plain, */*\r\n' +
    'Content-Type: application/json\r\n' +
    'User-Agent: axios/0.26.0\r\n' +
    'Content-Length: 2\r\n' +
    'Host: localhost:3333\r\n' +
    'Cookie: key=value\r\n' +
    'Connection: keep-alive\r\n' +
    '\r\n' +
    '{}POST /1 HTTP/1.1\r\n' +
    'Accept: application/json, text/plain, */*\r\n' +
    'Content-Type: application/json\r\n' +
    'User-Agent: axios/0.26.0\r\n' +
    'Content-Length: 2\r\n' +
    'Host: localhost:3333\r\n' +
    'Cookie: key=value\r\n' +
    'Connection: keep-alive\r\n' +
    '\r\n'
}
```

Note duplicate `POST /1 HTTP/1.1` parts - that means the client sent the request line and headers for one request twice. 

## Expected behavior

http-cookie-agent should work with keepAlive without sending invalid requests.

## Code that reproduces the bug

```typescript
import axios from 'axios'
import {HttpCookieAgent} from 'http-cookie-agent'
import {CookieJar} from 'tough-cookie'

// address for our test server.
const host = 'http://localhost:3333'

// arbitrary body. i've only seen this issue with POST requests with bodies. i
// think the http request must have something after the headers to trigger the
// error.
const body = {}

// the cookie jar must have a cookie for the host.
const jar = new CookieJar()
jar.setCookie('key=value', host)

// set keepAlive to true to demonstrate the issue. with keepAlive false i've
// never seen an error.
const httpAgent = new HttpCookieAgent({ jar, keepAlive: true, })
const client = axios.create({ httpAgent })

// send serial POST requests to our test server halting on the first error.
async function main() {
  let i = 0
  try {
    while (i < 100) {
      await client.post(`${host}/${i}`, body)
      i++
    }
    console.log('done')
  } catch (err: any) {
    console.log('error', i, err.message)
  }
}

main()
```
```typescript
// http server prints socket data to stdout and client errors to stderr.
import { Socket } from 'net'
import { createServer } from 'http'

const connections = new WeakSet<Socket>()
createServer((req, res) => {
  const socket = res.socket!

  console.error(`> ${req.method} ${req.url} ${connections.has(socket) ? 'reused' : ''}`)

  // pipe socket to stdout once per socket. sockets may be reused across requests.
  if (!connections.has(socket)) socket.pipe(process.stdout)
  connections.add(socket)

  res.end()
}).on('clientError', (err: any, socket) => {
  console.error(err.message, { packet: err.rawPacket.toString() })
  socket.end('HTTP/1.1 400 Bad Request\r\n\r\n');
})
.listen(3333, () => console.error('listening\n'))
```

## Environments

- OS: linux 5.15.0
- Node.js version: multiple (16.14.0, 14.19.0)
- axios version: 0.26.0
- tough-cookie version: 4.0.0
- http-cookie-agent: 1.0.4

**Comment by everett1992** [comment]
I haven't fully traced the code but what I'm seeing that that when `end` is called `_headerSent` is false so this block writes the _header to the socket, after header have been written when outputData is flushed.

https://github.com/nodejs/node/blob/0367b5c35ea0f98b323175a4aaa8e651af7a91e7/lib/_http_outgoing.js#L323-L339

Over this block req._headerSent becomes false, I don't know if it's setHeaders or _implicitHeaders that does it.

https://github.com/3846masa/http-cookie-agent/blob/8aa5ecb437029cef72568fb5154d1e9dabce772f/src/create_cookie_agent.ts#L102-L110 

I found that adding `req._headerSent = true` after when you edit the firstChunk fixes the error.

---

 I like the this libraries approach of supporting cookies with an Agent because it will support any http library that uses agents but this implementation uses a lot of private and not documented methods node's built ins. Its a hack and I'm worried it will be fragile. I have two thoughts for making the approach more sound
 
First, using [createConneciton](https://nodejs.org/api/http.html#agentcreateconnectionoptions-callback) you could wrap the socket with an http parser that injects cookies into the requests and extracts cookies from responses. createConnection is  public and parsing a tcp stream is a well defined operation. It will be slower.

Second, `addRequest` isn't  documented as far as I can tell, but it seems like a safe integration point if you can inject cookies sync. If you replaced tough-cookie with a sync cookie jar I think you could get away with setting cookies in `addRequest`.

The problem, as I see it, is as soon as you await other code can start writing data, flush headers, and end the stream before you get the cookie back. There will be a point when the headers have been flushed over the wire and there's no way to modify them. See `_writeRaw` which bypasses the outputData buffer.

**Comment by 3846masa** [comment]
Thank you for reporting the issue.

I have investigated and found that the issue was caused by `_implicitHeader()` causing the value of `_headerSent` to be set to `false`.

I am working on a fix for this issue in #90, which will be released as v1.0.6.

As you mentioned, this library relies on the private behavior of Node.js.

In the future I will take your advice and implement it in a more robust way.
